File: Packages/HeistEngine/Sources/HeistCore/Simulation/Vision.swift
import Foundation

public enum VisionSourceKind: String, Sendable, Codable, Equatable {
    case guardActor
    case securityCamera
}

/// The authored numbers that define a watcher. Kept in HeistCore so planning,
/// playback and the RealityKit visualization all consume the same values.
public struct VisionConfig: Sendable, Equatable {
    public var range: Double
    public var fieldOfViewDegrees: Double

    public init(range: Double, fieldOfViewDegrees: Double) {
        self.range = range
        self.fieldOfViewDegrees = fieldOfViewDegrees
    }
}

public enum VisionProfiles {
    /// Guards read farther but have a deliberately narrower human cone.
    public static let guardActor = VisionConfig(range: 10, fieldOfViewDegrees: 70)
    /// Cameras cover a short, broad slice of space.
    public static let securityCamera = VisionConfig(range: 6, fieldOfViewDegrees: 120)
}

/// Pure deterministic scan motion used by security cameras.
public enum VisionScan {
    public static func facing(
        baseDegrees: Double,
        arcDegrees: Double,
        period: Double,
        time: Double
    ) -> Double {
        guard arcDegrees != 0, period > 0 else {
            return PatrolRoute.normalizedYaw(baseDegrees)
        }
        let phase = (time / period) * 2 * Double.pi
        return PatrolRoute.normalizedYaw(baseDegrees + sin(phase) * arcDegrees / 2)
    }
}

/// Exact reason a target is or is not seen. Mission failure can therefore be
/// explained to the player instead of being reduced to an opaque Boolean.
public enum VisionResult: Sendable, Equatable {
    case visible(distance: Double)
    case outOfRange(distance: Double)
    case outsideFieldOfView(angleDegrees: Double)
    case occluded(by: String)

    public var isVisible: Bool {
        if case .visible = self { return true }
        return false
    }
}

public struct DetectionEvent: Sendable, Equatable {
    public var sourceID: String
    public var sourceKind: VisionSourceKind
    public var targetID: String
    public var missionTime: Double

    public init(
        sourceID: String,
        sourceKind: VisionSourceKind,
        targetID: String,
        missionTime: Double
    ) {
        self.sourceID = sourceID
        self.sourceKind = sourceKind
        self.targetID = targetID
        self.missionTime = missionTime
    }
}

/// Deterministic line of sight for guards and cameras. Range and facing are
/// measured on the floor plane; occlusion uses the full wall volume so a
/// doorway lintel does not behave like a solid wall at eye height.
public enum VisionSolver {
    public static func evaluate(
        observer: WorldPoint,
        facingDegrees: Double,
        target: WorldPoint,
        config: VisionConfig,
        occluders: [WorldBox]
    ) -> VisionResult {
        let offset = (target - observer).onFloorPlane
        let distance = offset.planarLength

        guard distance <= max(0, config.range) else {
            return .outOfRange(distance: distance)
        }

        if distance > 1e-9 {
            let yaw = facingDegrees * .pi / 180
            let forward = WorldPoint(x: sin(yaw), y: 0, z: cos(yaw))
            let cosine = max(-1, min(1, forward.dot(offset) / distance))
            let angle = acos(cosine) * 180 / .pi
            guard angle <= max(0, config.fieldOfViewDegrees) / 2 + 1e-9 else {
                return .outsideFieldOfView(angleDegrees: angle)
            }
        }

        // Stable source order makes the reported blocker deterministic even
        // when a ray crosses two wall segments at the same point.
        for wall in occluders.sorted(by: { $0.sourceID < $1.sourceID }) {
            if intersectionEntry(observer, target, box: wall) != nil {
                return .occluded(by: wall.sourceID)
            }
        }

        return .visible(distance: distance)
    }

    /// Distance a displayed sight ray may travel before the first occluder.
    /// The endpoint height lets a ceiling camera ray slope toward a person's
    /// eye height while a guard ray stays approximately horizontal.
    public static func visibleReach(
        observer: WorldPoint,
        targetHeight: Double,
        facingDegrees: Double,
        maxRange: Double,
        occluders: [WorldBox]
    ) -> Double {
        let range = max(0, maxRange)
        guard range > 0 else { return 0 }
        let yaw = facingDegrees * .pi / 180
        let end = WorldPoint(
            x: observer.x + sin(yaw) * range,
            y: targetHeight,
            z: observer.z + cos(yaw) * range
        )
        let firstEntry = occluders.compactMap {
            intersectionEntry(observer, end, box: $0)
        }.min() ?? 1
        return range * max(0, min(1, firstEntry))
    }

    private static func intersectionEntry(
        _ start: WorldPoint,
        _ end: WorldPoint,
        box: WorldBox
    ) -> Double? {
        let yaw = box.yaw * .pi / 180
        let c = cos(yaw), s = sin(yaw)

        func local(_ point: WorldPoint) -> (x: Double, z: Double) {
            let dx = point.x - box.center.x
            let dz = point.z - box.center.z
            return (x: c * dx - s * dz, z: s * dx + c * dz)
        }

        let a = local(start)
        let b = local(end)
        let direction = (x: b.x - a.x, z: b.z - a.z)
        let startY = start.y - box.center.y
        let deltaY = end.y - start.y
        var entry = 0.0
        var exit = 1.0

        func clip(origin: Double, delta: Double, halfExtent: Double) -> Bool {
            if abs(delta) < 1e-12 {
                return abs(origin) <= halfExtent
            }
            var near = (-halfExtent - origin) / delta
            var far = (halfExtent - origin) / delta
            if near > far { swap(&near, &far) }
            entry = max(entry, near)
            exit = min(exit, far)
            return entry <= exit
        }

        guard clip(origin: a.x, delta: direction.x, halfExtent: box.width / 2),
              clip(origin: a.z, delta: direction.z, halfExtent: box.depth / 2),
              clip(origin: startY, delta: deltaY, halfExtent: box.height / 2) else {
            return nil
        }

        // Contact exactly at either actor is not an intervening wall.
        return exit > 1e-9 && entry < 1 - 1e-9 ? max(0, entry) : nil
    }
}

File: UnityPort/Assets/DerClou/Core/Simulation/Vision.cs
namespace DerClou.Core.Simulation
{
    using System;
    using System.Collections.Generic;
    using DerClou.Core.Data;

    public enum VisionResultKind
    {
        Visible,
        OutOfRange,
        OutsideFieldOfView,
        Occluded
    }

    /// Exact, inspectable reason a target is or is not visible. Keeping this
    /// in Core makes a mission failure explainable instead of a physics-side
    /// Boolean that changes with render timing.
    public struct VisionResult
    {
        public VisionResultKind Kind;
        public float Distance;
        public float AngleDegrees;
        public string OccluderId;
        public bool IsVisible => Kind == VisionResultKind.Visible;
    }

    [Serializable]
    public class VisionSourceState
    {
        public int sourceId;
        public string sourceLabel;
        public int actorId;
        public VisionSourceKind kind;
        public VisionConfig config;
        public WorldPoint fixedPosition;
        public float eyeHeight = 1.55f;
        public float currentFacingYaw;
        public bool isEnabled = true;
    }

    [Serializable]
    public struct FailureEvent
    {
        public float time;
        public int actorId;
        public string source;
        public string reason;
        public WorldPoint position;
    }

    /// Pure deterministic floor-plane vision with footprint-only occlusion.
    /// No Physics.Raycast, Collider, Transform or frame delta participates.
    public static class VisionSolver
    {
        private const float Epsilon = 0.000001f;

        public static VisionResult Evaluate(
            WorldPoint observer,
            float facingDegrees,
            WorldPoint target,
            VisionConfig config,
            IReadOnlyList<WorldBox> occluders)
        {
            float dx = target.x - observer.x;
            float dz = target.z - observer.z;
            float distance = MathF.Sqrt(dx * dx + dz * dz);

            if (distance > MathF.Max(0f, config.range) + Epsilon)
            {
                return new VisionResult { Kind = VisionResultKind.OutOfRange, Distance = distance };
            }

            float angle = 0f;
            if (distance > Epsilon)
            {
                float yaw = facingDegrees * MathF.PI / 180f;
                float dot = (MathF.Sin(yaw) * dx + MathF.Cos(yaw) * dz) / distance;
                dot = MathF.Max(-1f, MathF.Min(1f, dot));
                angle = MathF.Acos(dot) * 180f / MathF.PI;
                if (angle > MathF.Max(0f, config.fieldOfViewDegrees) * 0.5f + Epsilon)
                {
                    return new VisionResult
                    {
                        Kind = VisionResultKind.OutsideFieldOfView,
                        Distance = distance,
                        AngleDegrees = angle
                    };
                }
            }

            if (occluders != null && occluders.Count > 0)
            {
                bool found = false;
                string stableId = null;
                for (int i = 0; i < occluders.Count; i++)
                {
                    var box = occluders[i];
                    if (!TryIntersectionEntry(observer, target, box, out _)) continue;
                    // Occluded is the gameplay fact. When several walls
                    // overlap the ray, expose a stable authored id for
                    // diagnostics regardless of input order, without sorting
                    // or allocating on every visibility check.
                    if (found && string.CompareOrdinal(box.sourceID, stableId) >= 0) continue;
                    found = true;
                    stableId = box.sourceID;
                }
                if (found)
                {
                    return new VisionResult
                    {
                        Kind = VisionResultKind.Occluded,
                        Distance = distance,
                        AngleDegrees = angle,
                        OccluderId = stableId
                    };
                }
            }

            return new VisionResult
            {
                Kind = VisionResultKind.Visible,
                Distance = distance,
                AngleDegrees = angle
            };
        }

        /// Distance a displayed cone ray reaches before the first simulation
        /// occluder. The presentation calls this same method, so the visible
        /// cone cannot disagree with detection geometry.
        public static float VisibleReach(
            WorldPoint observer,
            float targetHeight,
            float facingDegrees,
            float maxRange,
            IReadOnlyList<WorldBox> occluders)
        {
            float range = MathF.Max(0f, maxRange);
            if (range <= 0f) return 0f;

            float yaw = facingDegrees * MathF.PI / 180f;
            // Kept in the signature for source compatibility with existing
            // callers. Height is deliberately irrelevant to gameplay.
            _ = targetHeight;
            var end = new WorldPoint(
                observer.x + MathF.Sin(yaw) * range,
                0f,
                observer.z + MathF.Cos(yaw) * range);

            float firstEntry = 1f;
            if (occluders != null)
            {
                for (int i = 0; i < occluders.Count; i++)
                {
                    if (TryIntersectionEntry(observer, end, occluders[i], out float entry))
                        firstEntry = MathF.Min(firstEntry, entry);
                }
            }
            return range * MathF.Max(0f, MathF.Min(1f, firstEntry));
        }

        private static bool TryIntersectionEntry(
            WorldPoint start,
            WorldPoint end,
            WorldBox box,
            out float intersectionEntry)
        {
            float yaw = box.yaw * MathF.PI / 180f;
            float c = MathF.Cos(yaw), s = MathF.Sin(yaw);

            float startDx = start.x - box.centerX;
            float startDz = start.z - box.centerZ;
            float endDx = end.x - box.centerX;
            float endDz = end.z - box.centerZ;

            float ax = c * startDx - s * startDz;
            float az = s * startDx + c * startDz;
            float bx = c * endDx - s * endDz;
            float bz = s * endDx + c * endDz;

            float entry = 0f, exit = 1f;
            bool clipped = Clip(ax, bx - ax, box.width * 0.5f, ref entry, ref exit)
                && Clip(az, bz - az, box.depth * 0.5f, ref entry, ref exit);

            intersectionEntry = MathF.Max(0f, entry);
            return clipped && exit > Epsilon && entry < 1f - Epsilon;
        }

        private static bool Clip(float origin, float delta, float halfExtent, ref float entry, ref float exit)
        {
            if (MathF.Abs(delta) < Epsilon) return MathF.Abs(origin) <= halfExtent;
            float near = (-halfExtent - origin) / delta;
            float far = (halfExtent - origin) / delta;
            if (near > far) { float swap = near; near = far; far = swap; }
            entry = MathF.Max(entry, near);
            exit = MathF.Min(exit, far);
            return entry <= exit;
        }
    }
}
